### PR TITLE
Profiling and some small optimization

### PR DESCRIPTION
**Pull request by Vuwij**

### CODE REVIEW

**Comment by Vuwij** [comment]
Optimizations:
- Changed double to float
- RNG is now in the threadpool
- Split up the std::transforms seperately between intensities and ranges
- Basic profiler (tied to timekeeper). use START_PROFILE and END_PROFILE to  start and end the proflier

File: flatland_plugins/src/laser.cpp [hunk 1]
```diff
@@ -180,59 +180,52 @@ void Laser::ComputeLaserRanges() {
...
+        *i = result.second;
       }
     }
-
   } else {
     if (flipped_) {
       std::transform(results.begin(), results.end(),
-                     laser_scan_.ranges.rbegin(), range_trans);
+                     laser_scan_.ranges.rbegin(),
+                     [this](std::future<std::pair<float, float>>& res) {
```
Comment (hendrik-skubch) [inline flatland_plugins/src/laser.cpp]: no need to capture this here

File: flatland_plugins/src/laser.cpp [hunk 2]
```diff
@@ -180,59 +180,52 @@ void Laser::ComputeLaserRanges() {
...
+                       return res.get().first;
+                     });
     } else {
       std::transform(results.begin(), results.end(), laser_scan_.ranges.begin(),
-                     range_trans);
+                     [this](std::future<std::pair<float, float>>& res) {
```
Comment (hendrik-skubch) [inline flatland_plugins/src/laser.cpp]: and here

File: flatland_server/include/flatland_server/profiler.h [hunk 3]
```diff
@@ -0,0 +1,99 @@
+#ifndef FLATLAND_PROFILER_H
+#define FLATLAND_PROFILER_H
+
+#define PROFILER_ON false
+#define PROFILER_OUTPUT_PATH "/tmp/flatland_profile_output.log"
+
+#if PROFILER_ON == true
```
Comment (hendrik-skubch) [inline flatland_server/include/flatland_server/profiler.h]: #if PROFILER_ON

File: flatland_server/include/flatland_server/profiler.h [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+  void end() {
+    assert(started_);
+    lapse_count_++;
+    total_duration_ += std::chrono::duration_cast<std::chrono::microseconds>(
+        std::chrono::high_resolution_clock::now() - start_time_);
+  }
+
+  int getLapseCount() const { return lapse_count_; }
+
+  long getTotalDuration() const {
```
Comment (hendrik-skubch) [inline flatland_server/include/flatland_server/profiler.h]: long?
Prefer explicit types, so int32_t. Or, if that's what you wanted, int64_t

File: flatland_server/include/flatland_server/profiler.h [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+  std::chrono::high_resolution_clock::time_point start_time_;
+  std::chrono::high_resolution_clock::duration total_duration_;
+};
+
+class Profiler {
+ public:
+  Profile& get(const std::string& profile_name) {
+    auto p = profiles_.find(profile_name);
+    if (p == profiles_.end()) {
+      auto profile = profiles_.emplace(profile_name, Profile());
```
Comment (hendrik-skubch) [inline flatland_server/include/flatland_server/profiler.h]: profiles_.emplace(profile_name);

File: flatland_server/src/plugin_manager.cpp [hunk 6]
```diff
@@ -77,19 +77,39 @@ PluginManager::~PluginManager() {
...
     model_plugin->AfterPhysicsStep(timekeeper_);
+    END_PROFILE(timekeeper_, "After Physics Step: " +
+                                 model_plugin.get()->GetModel()->name_ + " " +
+                                 model_plugin.get()->name_);
   }
   for (const auto &world_plugin : world_plugins_) {
+    START_PROFILE(timekeeper_,
+                  "After Physics Step: " + world_plugin.get()->name_);
     world_plugin->AfterPhysicsStep(timekeeper_);
+    START_PROFILE(timekeeper_,
```
Comment (hendrik-skubch) [inline flatland_server/src/plugin_manager.cpp]: End Profile

File: flatland_server/include/flatland_server/profiler.h [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+ private:
+  int lapse_count_;
+  bool started_;
+  std::chrono::high_resolution_clock::time_point start_time_;
+  std::chrono::high_resolution_clock::duration total_duration_;
+};
+
+class Profiler {
+ public:
+  Profile& get(const std::string& profile_name) {
```
Comment (hendrik-skubch) [inline flatland_server/include/flatland_server/profiler.h:63]: I'm not a big fan of using strings...for anything. For now I think it is ok, if we want to measure more accurately we should replace them though.

**Review by hendrik-skubch — CHANGES_REQUESTED** [review]

File: flatland_server/include/flatland_server/profiler.h (hunk 4 above)
Comment (Vuwij) [reply flatland_server/include/flatland_server/profiler.h]: by default the function returns long. What is the advantages of using int64_t?

File: flatland_server/include/flatland_server/profiler.h (hunk 7 above)
Comment (Vuwij) [reply flatland_server/include/flatland_server/profiler.h:63]: ok

File: flatland_server/include/flatland_server/profiler.h (hunk 5 above)
Comment (Vuwij) [reply flatland_server/include/flatland_server/profiler.h]: This is an unordered_map. I don't think it compiles with above?

File: flatland_server/src/plugin_manager.cpp (hunk 6 above)
Comment (Vuwij) [reply flatland_server/src/plugin_manager.cpp]: thanks good catch

File: flatland_server/include/flatland_server/profiler.h (hunk 4 above)
Comment (hendrik-skubch) [reply flatland_server/include/flatland_server/profiler.h]: The function? Which function?
int64_t is bigger than a long, and a has fixed size.

File: flatland_server/include/flatland_server/profiler.h (hunk 5 above)
Comment (hendrik-skubch) [reply flatland_server/include/flatland_server/profiler.h]: you are correct, but this does:
`auto profile = profiles_.emplace(std::piecewise_construct,std::forward_as_tuple(profile_name), std::forward_as_tuple());`
:D

File: flatland_server/include/flatland_server/profiler.h (hunk 4 above)
Comment (Vuwij) [reply flatland_server/include/flatland_server/profiler.h]: Yeah i got confused with long and double

**Review by hendrik-skubch — APPROVED** [review]

**Pull request merged**